File: ruby/bob/_iterations/1/bob.rb
module Bob
  def self.hey(remark)
    return "Fine. Be that way!" if /\A\s*\Z/ =~ remark
    question = /\?\s*\Z/ =~ remark
    yelling = /\A[^a-z]*[A-Z][^a-z]*\Z/ =~ remark
    return "Calm down, I know what I'm doing!" if question && yelling
    return "Whoa, chill out!" if yelling
    return "Sure." if question
    "Whatever."
  end
end

File: clojure/bob/_iterations/2/src/bob.clj
(ns bob
  (:require [clojure.string :refer [trim ends-with?]]))

(defn- yell? [s]
  (and (re-find #"[A-Z]" s) (nil? (re-find #"[a-z]" s))))

(defn- question? [s]
  (ends-with? s "?"))

(defn response-for [s]
  (let [trimmed (trim s)
        is-yelling (yell? trimmed)
        is-question (question? trimmed)]
    (cond
      (and is-yelling is-question) "Calm down, I know what I'm doing!"
      is-yelling                   "Whoa, chill out!"
      is-question                  "Sure."
      (empty? trimmed)             "Fine. Be that way!"
      :else                        "Whatever.")))
